ee/management/commands/backfill_persons_and_groups_on_events.py

```diff
@@ -0,0 +1,191 @@
+LAYOUT(complex_key_cache(size_in_cells 1000))
+Lifetime(60000)
+"""
+
+backfill_settings = "SETTINGS mutations_sync = 2" if settings.TEST else ""
+
+BACKFILL_SQL = f"""
+ALTER TABLE {EVENTS_DATA_TABLE()}
+ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+UPDATE
+    person_id=toUUID(dictGet('{CLICKHOUSE_DATABASE}.{PERSON_DISTINCT_IDS_DICT_TABLE_NAME}', 'person_id', distinct_id)),
+    person_properties=dictGetString('{CLICKHOUSE_DATABASE}.{PERSONS_DICT_TABLE_NAME}', 'properties', toUUID(dictGet('{CLICKHOUSE_DATABASE}.{PERSON_DISTINCT_IDS_DICT_TABLE_NAME}', 'person_id', distinct_id))),
+    group0_properties=dictGetString('{CLICKHOUSE_DATABASE}.{GROUPS_DICT_TABLE_NAME}', 'group_properties', $group_0),
+    group1_properties=dictGetString('{CLICKHOUSE_DATABASE}.{GROUPS_DICT_TABLE_NAME}', 'group_properties', $group_1),
+    group2_properties=dictGetString('{CLICKHOUSE_DATABASE}.{GROUPS_DICT_TABLE_NAME}', 'group_properties', $group_2),
+    group3_properties=dictGetString('{CLICKHOUSE_DATABASE}.{GROUPS_DICT_TABLE_NAME}', 'group_properties', $group_3),
+    group4_properties=dictGetString('{CLICKHOUSE_DATABASE}.{GROUPS_DICT_TABLE_NAME}', 'group_properties', $group_4)
+WHERE team_id = %(team_id)s
+{backfill_settings}
+"""
+
+GET_QUERY_ID_SQL = f"""
+SELECT query_id
+FROM system.query_log
+WHERE
+    query_start_time > (now() - INTERVAL 10 MINUTE) AND
+    query LIKE '%backfill:{backfill_query_id}%' AND
+    query NOT LIKE '%query_log%'
+LIMIT 1
+"""
+
+DROP_GROUPS_DICTIONARY_SQL = f"DROP DICTIONARY IF EXISTS {GROUPS_DICT_TABLE_NAME}"
+DROP_PERSON_DISTINCT_IDS_DICTIONARY_SQL = f"DROP DICTIONARY IF EXISTS {PERSON_DISTINCT_IDS_DICT_TABLE_NAME}"
+DROP_PERSONS_DICTIONARY_SQL = f"DROP DICTIONARY IF EXISTS {PERSONS_DICT_TABLE_NAME}"
+
+query_number = 0
+
+
+def print_and_execute_query(sql: str, name: str, dry_run: bool, timeout=180, query_args={}) -> Any:
+    global query_number
+
+    if not settings.TEST:
+        print(f"> {query_number}. {name}", end="\n\n")
+        print(sql, end="\n")
+        print("---------------------------------", end="\n\n")
+
+    query_number = query_number + 1
+
+    if not dry_run:
+        res = sync_execute(sql, settings={"max_execution_time": timeout}, args=query_args)
+        return res
+
+    return None
+
+
+def run_backfill(options):
+
+    if not options["team_id"]:
+        logger.error("You must specify --team-id to run this script")
+        exit(1)
+
+    dry_run = not options["live_run"]
+
+    if dry_run:
+        print("Dry run. Queries to run:", end="\n\n")
+
+    print_and_execute_query(GROUPS_DICTIONARY_SQL, "GROUPS_DICTIONARY_SQL", dry_run)
+    print_and_execute_query(PERSON_DISTINCT_IDS_DICTIONARY_SQL, "PERSON_DISTINCT_IDS_DICTIONARY_SQL", dry_run)
+    print_and_execute_query(PERSONS_DICTIONARY_SQL, "PERSONS_DICTIONARY_SQL", dry_run)
+
+    client._request_information = {"kind": "backfill", "id": backfill_query_id}
+    print_and_execute_query(
```

Comment:
Will this block and wait for success - I don't think it will right?

Reply:
It shouldn't - notice how on tests I set https://clickhouse.com/docs/en/operations/settings/settings#mutations_sync.

An `ALTER TABLE` should just queue the mutations to be run async

Reply:
Then why can we drop the dictionaries after this returns?

Reply:
oof great point

Reply:
yeah maybe best to keep them around and only drop manually later - them being around is no problem either.

Reply:
I added the suffix to them but I think it could be useful to just use a standard name and reuse them?

Shouldn't add any overhead. I can reduce the cache size if we want. Altinity mentioned how this value could effectively be 0, shouldn't change too much - we'll just have the interface for pulling data